### ISSUE TITLE
poetry 1.2.0 can't install packages from private pypi servers supporting only MD5 hashes

### ISSUE THREAD

**Issue by jclerman** [python-poetry/poetry#6301]
- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) Poetry version.
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option).

- **OS version and name**: MacOS Monterey (12.5.1)
- **Poetry version**: 1.2.0

## Issue
After upgrading from poetry 1.1.15 to 1.2.0, I can no longer `poetry install` in my dev environment - it fails on a single package.  I've investigated and it **looks** like the problem is that with the new `poetry`, only SHA256 hashes are stored for packages in `poetry.lock`, even when the serving PyPi repository only supplies MD5 hashes.

Somehow, even though the PyPi repo serves MD5 (only) hashes, the new poetry stores a SHA256 hash in `poetry.lock` - then when I try to do `poetry install`, the MD5 hash from the retrieved package of course doesn't match the hash in the lock file, so I get:

```
Retrieved digest for link sirona-xyz-utils-0.10.2.tar.gz(md5:XXXXX) not in poetry.lock metadata ['sha256:YYYYY', 'sha256:ZZZZZ']

  at ~/Library/Application Support/pypoetry/venv/lib/python3.7/site-packages/poetry/installation/chooser.py:145 in _get_links
      141│             selected_links.append(link)
      142│
      143│         if links and not selected_links:
      144│             raise RuntimeError(
    → 145│                 f"Retrieved digest for link {link.filename}({h}) not in poetry.lock"
      146│                 f" metadata {hashes}"
      147│             )
      148│
      149│         return selected_links
```
(I've obscured the actual hashes and package-name above).

In our case, the PyPi repo server is Nexus, and not the very latest Nexus which does serve SHA256 hashes - that was only released this month, Aug 2022.

**Comment by jclerman** [comment]
I can work around the issue by manually installing the offending package(s) via `pip`, but that doesn't address the real issue here.

**Comment by neersighted** [comment]
Duplicate #4523

**Comment by jclerman** [comment]
@neersighted, is this really a duplicate?  The behavior I've noted here is a regression from 1.1.15 to 1.2.0.

**Comment by andrecp** [comment]
This isn't duplicated @neersighted , I think @dimbleby  also didn't read the latest comments. 1.2.0 broke poetry for a lot of its current users. downgrading works.

**Comment by dimbleby** [comment]
Well in that case we can close #4523 and leave this one open, because 1.1 is out of support so it's only the 1.2 behaviour that's worth tracking here.

Either way we only need one open issue.

For real, please stop tagging me: it was borderline rude the first time and it's absurd after I've asked you to stop.

**Comment by andrecp** [comment]
> Well in that case we can close #4523 and leave this one open, because 1.1 is out of support so it's only the 1.2 behaviour that's worth tracking here.
> 
> Either way we only need one open issue.
> 
> For real, please stop tagging me: it was borderline rude the first time and it's absurd after I've asked you to stop.

Apologies, you mentioned "don't tag me in this issue", and this is a separate issue from my eyes, thought you hadn't read the comments above it in the other issue (as the thread changed context) I won't be tagging you anymore regarding this.

**Comment by neersighted** [comment]
Fair enough, going with this issue as the new canon.

I would like to point out that @dimbleby is NOT a Poetry maintainer -- he is a frequent contributor and an expert in many of the hairier parts of the ecosystem and codebase. However, he has expressly declined to be declared formally part of the project, and should not be pinged on issues he has not explicitly expressed interest in.

**Comment by dimbleby** [comment]
>  ... that @dimbleby ...

omg now you're at it...!  I'm also not entirely sure about github tag etiquette and hope I don't sound too grumpy; but would y'all please allow me to unsubscribe from this problem?  It just happens not to be interesting to me, that's all.

**Comment by andrecp** [comment]
Sounds good, apologies again!

Getting back in this subject, from @msscaroso on the #4523 issue

> Same issue happening here.
> I'm using poetry-1.2.0
> 
> Poetry create the lock file with md5 and now it only looks for sha256. The package repository has both and the install should not fail.
> 
> I'll try to pin poetry version and see if it works for the time being

I wonder if this was the regression ? It seems like in the case above the repo has both md5 and sha256, but poetry is not looking for the sha256 one. Anyho, will get back to work. For now I'm pinned to the latest version of `1.1.x` and it "fixed" this problem.

**Comment by neersighted** [comment]
> omg now you're at it...! I'm also not entirely sure about github tag etiquette and hope I don't sound too grumpy; but would y'all please allow me to unsubscribe from this problem? It just happens not to be interesting to me, that's all.

😆 my bad, was meant to be a one-off. Please don't ping non-maintainers (and please don't passive-aggressively ping maintainers either, not that it happened in this thread) as it's rather rude to do so, and this issue tracker is actively monitored. We have a lot of work to do regarding duplicates, labeling, and whatnot, but that is neither here nor there right now.

**Comment by LibofRelax** [comment]
For anyone desperately looking for a solution. Hopefully this helps. I certainly got real frustrated by it.

On Macbook Pro M1 Pro
MacOS: Monterey

1. Uninstall poetry 1.2.0 and install 1.1.5
2. Go to poetry's own virtual env, in my case, `~/Library/Application Support/pypoetry/venv/bin`
3. use `./pip list` to check `poetry-core`'s version. I got 1.0.8
4. `./pip install poetry-core==1.0.4` to downgrade it

Downgrading `poetry-core` is due to it hardcodes sha256 when validating hashes. We'll need to reverse that.

**Comment by DonFortes** [comment]
The same problem. pip install "poetry<1.2" as a temporary "fix" worked

**Comment by ronlut** [comment]
I have a similar problem:
lock file contains sha256 hash, but still poetry fails on the md5 has. Where does it get it from? I removed all cached and cache/artifact directories

**Comment by lucaspolo** [comment]
Same problem here! Temporary solution is pin version to <1.2

**Comment by lokeshjain279** [comment]
I am also facing the same problem and this is a major blocker in adopting poetry. Getting following error while installing a privately hosted library in nexus.
`
Retrieved digest for link mal_lib_poc-0.5.2.tar.gz(md5:0d08bdd0f8a21cb012816e77cec1d9c1) not in poetry.lock metadata ['sha256:605264c4a71648bbf4d966bc3031ac4121b1ecfea5ae730fe46257a36857233f', 'sha256:c021d5b2c34b68aeecb813a9108ddf3d9b19e2ebb1a93b4cb24a489c80417dd9']

  at ~/.local/share/pypoetry/venv/lib/python3.7/site-packages/poetry/installation/chooser.py:145 in _get_links
      141│             selected_links.append(link)
      142│ 
      143│         if links and not selected_links:
      144│             raise RuntimeError(
    → 145│                 f"Retrieved digest for link {link.filename}({h}) not in poetry.lock"
      146│                 f" metadata {hashes}"
      147│             )
      148│ 
      149│         return selected_links
`

**Comment by jclerman** [comment]
For people suffering from this issue and using the latest (3.41.x or above) version of Sonatype Nexus for their private PyPi repository, the problem can be solved by forcing Nexus to re-create the index files for all affected packages.  There is a Nexus bug where the problem is described: https://issues.sonatype.org/browse/NEXUS-34950 and the solution is to delete all files under `simple/` in the Nexus PyPi server (or at least, all files for packages that aren't yielding sha256 hashes yet).  The files will be re-created properly next time a client requests package info via the simple interface.

**Comment by shea-parkes** [comment]
I did find a stupid work around that seems to work for those using services like GemFury/Nexus that are still serving some packages with MD5 hashes only:

You can't just replace the MD5 hash with the SHA256 hash in `poetry.lock` because Poetry still compares the reported hash from the repository to `poetry.lock`  (and then goes on to hash the downloaded file with SHA256 and compare that against `poetry.lock` as well).

But at the moment, it appears you can have redundant file entries in `poetry.lock`.  So instead of replacing the hash for a given `.whl` (or whatever file you're failing on), duplicate that line and change the hash on the new line.  So you have two lines describing the same file, but one has an `md5` hash and one has a `sha256` hash.

The poetry internals appear to currently build a set of all hashes for a given package, so it's lenient on having multiple entries (as of current version of `poetry`).

**Comment by invokermain** [comment]
> The solution is to delete all files under `simple/` in the Nexus PyPi server.

I think this will fix the issue I am seeing with this poetry issue and Sonatype Nexus Repository ManagerOSS `3.37.1-0`, but I cannot figure out how to apply the fix. Could you ELI5?

**Comment by vad** [comment]
@invokermain you must upgrade Nexus to 3.41+ first, then you can:

- push new versions of the packages
- use that trick (remove `simple/`) to force sha256 hash creation

**Comment by AdamGagorik** [comment]
I see a similar issue for [packagecloud.io](https://packagecloud.io/) based private package repo.  However, it is with SHA1 hashes instead of MD5.  The solution provided by @shea-parkes works to get around it temporarily.  Manually editing the lock file every time we do a dependency update would not work long term.

```bash
• Installing xxxxxx (1.0.0)
 
RuntimeError
 
Retrieved digest for link xxxxxx-1.0.0.tar(sha1:...) not in poetry.lock metadata ['sha256:...']
 
at ~/.local/lib/python3.9/site-packages/poetry/installation/chooser.py:144 in _get_links
    140│ 
    141│             selected_links.append(link)
    142│ 
    143│         if links and not selected_links:
  → 144│             raise RuntimeError(
    145│                 f"Retrieved digest for link {link.filename}({h}) not in poetry.lock"
    146│                 f" metadata {hashes}"
    147│             )
    148│ 
```

**Comment by oleks-popovych** [comment]
Currently I'm facing the same issue.
Possible solution I see now - upgrading Nexus to recent version, which is finally switched from md5 to sha256.

This [issue] (https://issues.sonatype.org/browse/NEXUS-24127) may clarify why this issue appears.

In other hand, maybe _poetry_ could make step further to check which hash name repository service is providing?

**Comment by oleks-popovych** [comment]
updating Nexus to 3.41.1 and re-uploading packages worked for me.
but weirdly I've needed to re-upload (meaning downloading artifact, deleting it in nexus and uploading it again) changed hash to sha256 for all other versions of a package

**Comment by rykov** [comment]
@shea-parkes Michael from @gemfury here. We support both SHA256 and MD5 checksums. Newer accounts default to SHA256, and we have recently added a switch to migrate older accounts. If you're seeing MD5 checksums, we recommend heading to Settings/Features and migrating to SHA256.

**Comment by shea-parkes** [comment]
> @shea-parkes Michael from @gemfury here. We support both SHA256 and MD5 checksums. Newer accounts default to SHA256, and we have recently added a switch to migrate older accounts. If you're seeing MD5 checksums, we recommend heading to Settings/Features and migrating to SHA256.

Thank you very much sir.  I went to the account settings and found that switch.  It worked nicely.

I had opened a Gemfury support ticket on Sept 6th asking about switching to all SHA256 hashes (sorry I didn't find the self-service migration switch).  I just checked my inbox just now and still don't see a response.  Would you be able to nudge the support department to respond and close that one down?  (My Gemfury account is `shea-parkes` as well).

**Comment by marianhlavac** [comment]
&lt;vent&gt; I'm seriously being more and more angry with Poetry, last few months the tool is unusable and I'm seriously considering moving all projects to different package manager. I was waiting for 1.2.0 to solve all our issues with this tool, and while it did solve those issues, it created multiple new issues, which, again, prevents us to install our projects. I'm really sad, and I like Poetry a lot. &lt;/vent&gt;

**Comment by marianhlavac** [comment]
Some of you using pypiserver might find this as a solution (it did get lost a bit in the hidden comments in the related issue):

https://github.com/python-poetry/poetry/issues/4523#issuecomment-923741164

> For people who are using [pypiserver](https://pypi.org/project/pypiserver/) you can update the --hash-algo without any side effects (at least not on my side).

**Comment by jmuddle-via** [comment]
>  Michael from @gemfury here. We support both SHA256 and MD5 checksums. Newer accounts default to SHA256, and we have recently added a switch to migrate older accounts. If you're seeing MD5 checksums, we recommend heading to Settings/Features and migrating to SHA256.

Hi Michael @rykov, thank you for the update. What are the implications of making this switch? I assume we would have to update our `poetry.lock` files.

**Comment by strue36** [comment]
It looks like this is an issue that was fixed on the 1.1.15 branch by this commit 5221df927a8d46047ab705d5f42128c30ba1bd99. Would it be possible to port this change to the 1.2.x branch?

I believe that https://github.com/python-poetry/poetry/pull/5326 attempted to incorporate this but is still open.

I also found a branch by @neersighted (neersighted/hash_algos) that incorporates this change.

**Comment by volker48** [comment]
This issue doesn't seem to be related only to private pypi servers supporting only MD5 hashes as I have a private artifactory repo that has md5, sha1, and sha256 hashes and poetry 1.2.0 is still refusing to install the packages. This was working before the 1.2.0 update as well. 

It seems so far that this is only happening for internal private packages that are not published to pypi.

**Comment by sjlutterbie** [comment]
There once was a tool called poetry
Where deps could have a hash, two, or three.
But with the right call,
Just one can install -
Tough luck for your dependency tree.

**Comment by bsnacks000** [comment]
> @shea-parkes Michael from @gemfury here. We support both SHA256 and MD5 checksums. Newer accounts default to SHA256, and we have recently added a switch to migrate older accounts. If you're seeing MD5 checksums, we recommend heading to Settings/Features and migrating to SHA256.

Thanks so much for posting this here... I would have never found it... our CI was inexplicably broken for 3 hours today and I was going crazy reading these threads ... :beers:

**Comment by rykov** [comment]
> What are the implications of making this switch? I assume we would have to update our `poetry.lock` files.

Yes, you'll need to update the checksums in your `poetry.lock`, otherwise you're risking failed installations.

**Comment by KyleKing** [comment]
FWIW, we ran into an issue that after upgrading Nexus, private packages still had the md5 checksum: `Retrieved digest for link <package_name>-<ver>-py3-none-any.whl(md5:...) not in poetry.lock metadata ['sha256:...']` from API response of `<nexus_url>/repository/<repository_name>/simple/<internal_package_name>` (each href in the response text includes `#md5:...`)

This is a known issue for packages uploaded pre-Nexus upgrade: https://issues.sonatype.org/browse/NEXUS-34950 and the only resolution seems to be to re-upload the files (curl, delete, publish) or release new incremental versions

**Comment by limedaniel** [comment]
@KyleKing how exactly do you re-upload files though?  Presumably `pypiserver` will prevent you from trying to upload a package via `twine` or something.  Do you have some sample shell commands to share?

**Comment by KyleKing** [comment]
I thought there might have been a way to upload and replace, but we found it easier to release patch versions because we prevent replacing releases

**Comment by NickG123** [comment]
FYI - I had success with working around this issue on Poetry 1.2 by disabling the experimental new installer
```
poetry config experimental.new-installer false
```
for those stuck on a repository without support but don't want to downgrade

**Comment by zen-xu** [comment]
> FYI - I had success with working around this issue on Poetry 1.2 by disabling the experimental new installer
> 
> ```
> poetry config experimental.new-installer false
> ```
> 
> for those stuck on a repository without support but don't want to downgrade

If the `new-installer` is an experimental feature, please **don't enable by default** 🙏

**Comment by neersighted** [comment]
The new installer has been stable for several years, and disabling it is the experimental option. Doing so is under-tested and deviates from documented behavior.

**Comment by dimbleby** [comment]
On the plus side, the number of repositories that return md5 hashes is ever-dwindling

- artifactory https://www.jfrog.com/jira/browse/RTFACT-18495
- nexus https://issues.sonatype.org/browse/NEXUS-24127
- gemfury per a comment in this thread https://github.com/python-poetry/poetry/issues/6301#issuecomment-1241416832

I understand that in some organisations it takes a while to update these things, and probably there are others that haven't yet caught up.  

So while I'm not exactly advocating for "do nothing" here: it is true that as time passes, that becomes a more reasonable choice.

**Comment by dimbleby** [comment]
sure, that's known and already covered in this thread.  But that too will solve itself, eventually, either by nexus fixing it or by old packages being replaced.

If a great MR shows up then I don't see why it wouldn't be merged.  (And I'll note that more or less everyone who is interested in this is necessarily at least some sort of python developer...)  

But including #4523 this has been open for quite some time now: and that hasn't happened - and the pool of people who are affected is only shrinking as sha256 hashes becomes more and more available.

**Comment by aaditkamat** [comment]
> FYI - I had success with working around this issue on Poetry 1.2 by disabling the experimental new installer
> 
> ```
> poetry config experimental.new-installer false
> ```
> 
> for those stuck on a repository without support but don't want to downgrade

This worked for me but yeah it's not recommended as @neersighted has mentioned. The naming of the option as `experimental.new-installer` is kind of confusing though, because even I assumed (like @zen-xu) that the new installer is an experimental feature instead of disabling the new installer being an experimental option.

**Comment by waketzheng** [comment]
For pypi-server, start with flag `--hash-algo=sha256` also fix the error.
```bash
pypi-server run . --hash-algo=sha256
```

**Comment by limedaniel** [comment]
> `pypi-server run . --hash-algo=sha256`

This didn't work for us, as setting this only tells the server to generate a `sha256` hash for all uploads, *not for the packages already uploaded*.  I'm unaware of any means of telling the server to re-generate hashes using a new algorithm.

**Comment by ademoverflow** [comment]
Hello, same problem with latest poetry version (1.3.2).

We have a nexus that hosts home made python packages, and installing them with poetry does not work.

Logs:

```
Package operations: 1 install, 0 updates, 0 removals

  • Installing XXXXXXXXXXX (0.1.0): Failed

  RuntimeError

  Retrieved digest for link XXXXXXXXXXXX-0.1.0.tar.gz(md5:71cf85a955df5595ca799ee43dc470d3) not in poetry.lock metadata ['sha256:808761b60f05670f43c6be57d5c95ddc4db636041e62e74463a398c5cc9e4fb1', 'sha256:67ffb7d2f0a168517876ae8b98e075986d04c003419aec2543fc9f92c31184ea']

  at /opt/poetry/venv/lib/python3.8/site-packages/poetry/installation/chooser.py:146 in _get_links
      142│ 
      143│             selected_links.append(link)
      144│ 
      145│         if links and not selected_links:
    → 146│             raise RuntimeError(
      147│                 f"Retrieved digest for link {link.filename}({h}) not in poetry.lock"
      148│                 f" metadata {hashes}"
      149│             )
      150│ 
```

Cleared cache (both poetry and pip caches), remove poetry.lock, not working at all :(

I saw that this issue and another strongly similar exist since 3 years now, maybe this should be considered as a priority ?

**Comment by RajendraP** [comment]
> Hello, same problem with latest poetry version (1.3.2).
> 
> We have a nexus that hosts home made python packages, and installing them with poetry does not work.
> 
> Logs:
> 
> ```
> Package operations: 1 install, 0 updates, 0 removals
> 
>   • Installing XXXXXXXXXXX (0.1.0): Failed
> 
>   RuntimeError
> 
>   Retrieved digest for link XXXXXXXXXXXX-0.1.0.tar.gz(md5:71cf85a955df5595ca799ee43dc470d3) not in poetry.lock metadata ['sha256:808761b60f05670f43c6be57d5c95ddc4db636041e62e74463a398c5cc9e4fb1', 'sha256:67ffb7d2f0a168517876ae8b98e075986d04c003419aec2543fc9f92c31184ea']
> 
>   at /opt/poetry/venv/lib/python3.8/site-packages/poetry/installation/chooser.py:146 in _get_links
>       142│ 
>       143│             selected_links.append(link)
>       144│ 
>       145│         if links and not selected_links:
>     → 146│             raise RuntimeError(
>       147│                 f"Retrieved digest for link {link.filename}({h}) not in poetry.lock"
>       148│                 f" metadata {hashes}"
>       149│             )
>       150│ 
> ```
> 
> Cleared cache (both poetry and pip caches), remove poetry.lock, not working at all :(
> 
> I saw that this issue and another strongly similar exist since 3 years now, maybe this should be considered as a priority ?

We are also facing this issue :(

**Comment by ljnsn** [comment]
> Hello, same problem with latest poetry version (1.3.2).
> 
> We have a nexus that hosts home made python packages, and installing them with poetry does not work.
> 
> Logs:
> 
> ```
> Package operations: 1 install, 0 updates, 0 removals
> 
>   • Installing XXXXXXXXXXX (0.1.0): Failed
> 
>   RuntimeError
> 
>   Retrieved digest for link XXXXXXXXXXXX-0.1.0.tar.gz(md5:71cf85a955df5595ca799ee43dc470d3) not in poetry.lock metadata ['sha256:808761b60f05670f43c6be57d5c95ddc4db636041e62e74463a398c5cc9e4fb1', 'sha256:67ffb7d2f0a168517876ae8b98e075986d04c003419aec2543fc9f92c31184ea']
> 
>   at /opt/poetry/venv/lib/python3.8/site-packages/poetry/installation/chooser.py:146 in _get_links
>       142│ 
>       143│             selected_links.append(link)
>       144│ 
>       145│         if links and not selected_links:
>     → 146│             raise RuntimeError(
>       147│                 f"Retrieved digest for link {link.filename}({h}) not in poetry.lock"
>       148│                 f" metadata {hashes}"
>       149│             )
>       150│ 
> ```
> 
> Cleared cache (both poetry and pip caches), remove poetry.lock, not working at all :(
> 
> I saw that this issue and another strongly similar exist since 3 years now, maybe this should be considered as a priority ?

What version of Nexus are you running? It works fine since https://issues.sonatype.org/browse/NEXUS-24127, but new versions of the packages need to be uploaded.

**Comment by vmgustavo** [comment]
> pypi-server run . --hash-algo=sha256

this solved it for me

**Comment by ademoverflow** [comment]
> > Hello, same problem with latest poetry version (1.3.2).
> > We have a nexus that hosts home made python packages, and installing them with poetry does not work.
> > Logs:
> > ```
> > Package operations: 1 install, 0 updates, 0 removals
> > 
> >   • Installing XXXXXXXXXXX (0.1.0): Failed
> > 
> >   RuntimeError
> > 
> >   Retrieved digest for link XXXXXXXXXXXX-0.1.0.tar.gz(md5:71cf85a955df5595ca799ee43dc470d3) not in poetry.lock metadata ['sha256:808761b60f05670f43c6be57d5c95ddc4db636041e62e74463a398c5cc9e4fb1', 'sha256:67ffb7d2f0a168517876ae8b98e075986d04c003419aec2543fc9f92c31184ea']
> > 
> >   at /opt/poetry/venv/lib/python3.8/site-packages/poetry/installation/chooser.py:146 in _get_links
> >       142│ 
> >       143│             selected_links.append(link)
> >       144│ 
> >       145│         if links and not selected_links:
> >     → 146│             raise RuntimeError(
> >       147│                 f"Retrieved digest for link {link.filename}({h}) not in poetry.lock"
> >       148│                 f" metadata {hashes}"
> >       149│             )
> >       150│ 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Cleared cache (both poetry and pip caches), remove poetry.lock, not working at all :(
> > I saw that this issue and another strongly similar exist since 3 years now, maybe this should be considered as a priority ?
> 
> What version of Nexus are you running? It works fine since https://issues.sonatype.org/browse/NEXUS-24127, but new versions of the packages need to be uploaded.

I´m using  Nexus v3.37, but how this is related to Nexus ?

**Comment by ademoverflow** [comment]
> > I´m using Nexus v3.37, but how this is related to Nexus ?
> > @ademoverflow
> 
> Because Nexus prior to 3.41 doesn't report sha256 hashes, it only reports md5. AFAICT, Poetry 1.2+ does not support md5 hashes. You should upgrade to Nexus 3.44 if you need poetry 1.2+ or stick to poetry 1.1.x.

Thank you for the info. Will try to update the nexus asap.

But when I look at my package on Nexus, I see that it has some checksums, not only md5 (sha1, sha256, etc), so are we sure that the issue is really related to Nexus ONLY ?

**Comment by vfazio** [comment]
> > > I´m using Nexus v3.37, but how this is related to Nexus ?
> > > @ademoverflow
> > 
> > 
> > Because Nexus prior to 3.41 doesn't report sha256 hashes, it only reports md5. AFAICT, Poetry 1.2+ does not support md5 hashes. You should upgrade to Nexus 3.44 if you need poetry 1.2+ or stick to poetry 1.1.x.
> 
> Thank you for the info. Will try to update the nexus asap.
> 
> But when I look at my package on Nexus, I see that it has some checksums, not only md5 (sha1, sha256, etc), so are we sure that the issue is really related to Nexus ONLY ?

@ademoverflow 

Nexus stores multiple hashes for the individual backing files, but the page used to serve those files up as a pypi index only serves up the md5sum prior to 3.41. This is technically within the PEP 503 spec https://peps.python.org/pep-0503/ however poetry does not accept what they consider to be insecure hash types.

Nexus 3.44 supposedly fixes an issue with stale indices for packages uploaded prior to 3.41, but i haven't had a chance to test that it works.

**Comment by ademoverflow** [comment]
> > > > I´m using Nexus v3.37, but how this is related to Nexus ?
> > > > @ademoverflow
> > > 
> > > 
> > > Because Nexus prior to 3.41 doesn't report sha256 hashes, it only reports md5. AFAICT, Poetry 1.2+ does not support md5 hashes. You should upgrade to Nexus 3.44 if you need poetry 1.2+ or stick to poetry 1.1.x.
> > 
> > 
> > Thank you for the info. Will try to update the nexus asap.
> > But when I look at my package on Nexus, I see that it has some checksums, not only md5 (sha1, sha256, etc), so are we sure that the issue is really related to Nexus ONLY ?
> 
> @ademoverflow
> 
> Nexus stores multiple hashes for the individual backing files, but the page used to serve those files up as a pypi index only serves up the md5sum prior to 3.41. This is technically within the PEP 503 spec https://peps.python.org/pep-0503/ however poetry does not accept what they consider to be insecure hash types.
> 
> Nexus 3.44 supposedly fixes an issue with stale indices for packages uploaded prior to 3.41, but i haven't had a chance to test that it works.

After upgrading to 3.46 and reuploading my packages, it works perfectly. Thanks !

**Comment by ademoverflow** [comment]
> @ademoverflow
> 
> > After upgrading to 3.46 and reuploading my packages, it works perfectly. Thanks !
> 
> I don't want to take this too far off topic, but was reuploading necessary? Nexus documentation via https://issues.sonatype.org/browse/NEXUS-34950 implies this isn't necessary after 3.44 :-/

I've tested before reuploading with no success unfortunately. But right after the reupload, poetry succeed to add the package without any error.

**Comment by TreeKat71** [comment]
It does work for me.

Steps below
1. Update Nexus to `3.44`
2. Re-upload all needed packages.
3. Update poetry to the latest (`1.3.2`)
4. Reduce the time of `resolving dependencies` from 10000+s to 150s.

**Comment by vfazio** [comment]
Just for clarification, when updating Nexus to 3.44+ it's not necessary to reupload all packages. There are maintenance tasks that you can run to clear and generate the simple index https://issues.sonatype.org/browse/NEXUS-36608.

**Comment by palmitoto** [comment]
> > `pypi-server run . --hash-algo=sha256`
> 
> This didn't work for us, as setting this only tells the server to generate a `sha256` hash for all uploads, _not for the packages already uploaded_. I'm unaware of any means of telling the server to re-generate hashes using a new algorithm.

For my part it worked for all packages, old or new.

**Comment by falknerdominik** [comment]
> 

Also only worked for new packages for me. Maybe the nexus version was not high enough. In my case i reuploaded the packages.

**Comment by obaltian** [comment]
Since disabling new installer is now [deprecated (since 1.4.0)](https://github.com/python-poetry/poetry/blob/master/CHANGELOG.md#added), previous workaround should be replaced by something more solid.
Any updated on this issue?

**Comment by gaborbernat** [comment]
So the problem comes from https://github.com/python-poetry/poetry/blame/master/src/poetry/repositories/http_repository.py#L224-L246, change made by @abn; this ignores the hash returned by the HTTP repository and instead calculates SHA256 variant for it. However, at https://github.com/python-poetry/poetry/blob/master/src/poetry/installation/chooser.py#LL105C1-L105C1 this transformation is not made, hence the failure here. Options:

1. Respect the hash served by the server and do not recalculate.
2. Make the same dynamic calculation in `chooser`.

Maintainers, should I make a PR preferences?

**Comment by radoering** [comment]
I'm not that deep into this topic. After reading several linked issues and PRs, I conclude as follows regarding the two options:

**1. Respect the hash served by the server and do not recalculate.**

#2958 introduced the change and probably did not allow md5 because it's too weak (i.e. not secure). Further, this may result in other issues if the server is just a mirror, see https://github.com/python-poetry/poetry/pull/5326#issuecomment-1070927042:

> Allowing non-sha checksums would break exporting to requirements.txt for use with pip (unless using --without-hashes) as pip no longer supports using md5, sha1, and sha224 checksums.

[pip says](https://pip.pypa.io/en/stable/topics/secure-installs/#hash-algorithms):

> However, weaker ones such as md5, sha1, and sha224 are excluded to avoid giving a false sense of security.

I would consider simply allowing md5 to be a security risk. I think a config option to allow md5 that is turned off by default and maybe still prints a warning if md5 is used when turned on might be a viable option.

**2. Make the same dynamic calculation in chooser.**

Sounds like a "simple" bugfix to me so I'd say this might make more sense. Further, it's more secure. IMO, we should prefer this solution.

**Comment by lucas-bunzel** [comment]
@gaborbernat Do we have an estimate of when a release with this fix will be launched? Thank you.

**Comment by gaborbernat** [comment]
I'm not a maintainer so can't say.

**Comment by LionelR** [comment]
Hi all,
we use an old Artifactory server and this problem occurs with Poetry 1.5.1. Is there any plan to solve this issue? Thanks

**Comment by yiftachn** [comment]
Joining to ask for a solution, I can't affect the private pypi server owners

**Comment by chrimaho** [comment]
Bumping this issue. Is there any resolution to this problem yet?

**Comment by Jmennius** [comment]
Since Nexus's Jira is not available anymore (https://issues.sonatype.org/browse/NEXUS-24127 is not accessible)
and it was not easy to find adequate instruction, here is a brief summary:
* Upgrade to `Nexus >= 3.41.0`, it will support `sha256`.
  You will still have `md5` metadata for existing package, but you **do not have to re-upload** all your packages.
* Instead, in Nexus admin UI create a task `PyPI - Delete Index Asset MD5 Metadata` and run it once/manually for your repositories.

BTW, it looks like if you upload a new package - older ones will be automatically updated to use sha256. But I would still run the task manually.

**Comment by elementary-charge** [comment]
I'm using Nexus 3.43 and cannot just simply update that. For me helps a bit trivial

`poetry install --no-cache`

Previously I removed lock file too, if matters.
